xds/client.go

```diff
@@ -0,0 +1,199 @@
+package xds
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3endpointpb "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	v3discoverypb "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/golang/protobuf/proto"
+	statuspb "google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/gojek/courier-go/xds/backoff"
+	"github.com/gojek/courier-go/xds/log"
+)
+
+// stream is used to mock reloadingStream
+type stream interface {
+	Send(req *v3discoverypb.DiscoveryRequest) error
+	Recv() (*v3discoverypb.DiscoveryResponse, error)
+	startReloader(ctx context.Context)
+}
+
+// Options specifies options to be provided for initialising the xds client
+type Options struct {
+	XDSTarget       string
+	NodeProto       *v3corepb.Node
+	ClientConn      grpc.ClientConnInterface
+	BackoffStrategy backoff.Strategy
+	Logger          log.Logger
+}
+
+// NewClient returns a new ADS client stream using the *grpc.ClientConn provided.
+func NewClient(opts Options) (*Client, error) {
+	opts = setDefaultOpts(opts)
+
+	c := &Client{
+		xdsTarget:   opts.XDSTarget,
+		cc:          opts.ClientConn,
+		nodeProto:   opts.NodeProto,
+		done:        make(chan struct{}),
+		receiveChan: make(chan []*v3endpointpb.ClusterLoadAssignment),
+		log:         opts.Logger,
+	}
+
+	cc := opts.ClientConn
+	rs := &reloadingStream{
+		reloadCh:    make(chan error, 1),
+		onReConnect: c.onStreamReConnect,
+		connTimeout: 10 * time.Second,
+		log:         opts.Logger,
+		strategy:    opts.BackoffStrategy,
+		ns: func() v3discoverypb.AggregatedDiscoveryServiceClient {
```

Comment:
@ajatprabha, is ns a function to allow us to test better?

Reply:
yes, and it also takes away the need to inject grpc.ClientConnection from reloadingStream